README
This is a example github Python project. It is a template to allow easy creation of a new project - just copy this and amend as necessary.


CHANGES
0.1 Wed Mar 10 13:00:00 BST 2010

    * Initial commit

0.2 Mon Apr 5 10:30:00 BST 2010

    * Consideral work on SGM module
    * Updated ABresults.csv to conform to JSF2003IDresults.csv format

test/test_sgm.py
"""
strmarker test module.
"""

import unittest

from strprofiles import strmarker

class exampleTestCase(unittest.TestCase):

    def setUp(self):
	# AB Cau FGA frequencies (n=200)
	self.AB_Cau_FGA_alleles = {'18': 0.015, '19': 0.0625, '20': 0.1625, '20.2': 0.0075, '21': 0.1775, '22': 0.165, '22.2': 0.005, '23': 0.14, '24': 0.1325, '25': 0.1125, '26': 0.015, '27': 0.005}
	# AB Cau TH01 frequencies (n=200)
	self.AB_Cau_TH01_alleles = self.fromPercentages({'6':25.25, '7':16.00, '8':9.25, '9':13.75, '9.3':35.00, 10:0.75})
	# AB Cau D16S539 frequencies (n=200)
	self.AB_Cau_D16S539_alleles = self.fromPercentages({'8':1.5,'9':10.75,'10':4.25,'11':29.75,'12':34.25,'13':17.5,'14':1.75,'15':0.25})


    def tearDown(self):
        pass

    def testMarkerRandomMatchProbability(self):
        """marker random match probability"""

	# AB Cau FGA frequencies (n=200)
	alleles = self.AB_Cau_FGA_alleles
	expected = 0.036
	result = strmarker.calc_marker_rmp(alleles,0.0)
        self.assertAlmostEqual(result,expected,3)

	# AB Cau TH01 frequencies (n=200)
	alleles = self.AB_Cau_TH01_alleles
	expected = 0.094
	result = strmarker.calc_marker_rmp(alleles,0.0)
        self.assertAlmostEqual(result,expected,3)

	# AB Cau D16S539 frequencies (n=200)
	alleles = self.AB_Cau_D16S539_alleles
	expected = 0.103
	result = strmarker.calc_marker_rmp(alleles,0.0)
        self.assertAlmostEqual(result,expected,3)


    def fromPercentages(self,alleles):
        for i in alleles:
            alleles[i] /= 100.0
        return alleles


    def testGetModalProfile(self):
        """get modal profile"""
	item = {'name':'AB','marker':'VWA','alleles':{'5':0.94,'6':0.03,'7':0.02,'8':0.01}}
	data = [{'name':'AB','count':400,'marker':'FGA','alleles':self.AB_Cau_FGA_alleles}, \
		{'name':'AB','count':400,'marker':'TH01','alleles':self.AB_Cau_TH01_alleles}, \
		{'name':'AB','count':400,'marker':'D16S539','alleles':self.AB_Cau_D16S539_alleles},item]
	expected = {'FGA':(('21',0.1775),('22',0.165)), 'TH01':(('9.3',0.35),('6',0.2525)),'D16S539':(('12',0.3425),('11',0.2975)),'VWA':(('5',0.94),('5',0.94))}
	profile = expected
	result = strmarker.get_modal_profile(data,'AB')
	for i in expected:
		self.assertEqual(result[i][0][0],expected[i][0][0])
		self.assertEqual(result[i][1][0],expected[i][1][0])
		self.assertAlmostEqual(result[i][0][1],expected[i][0][1])
		self.assertAlmostEqual(result[i][1][1],expected[i][1][1])
	expected = 2*0.1775*0.165 * 2*0.35*0.2525 * 2*0.3425*0.2975 * 0.94*0.94
	result = strmarker.calc_profile_match_probability(profile,0.0)
	self.assertAlmostEqual(result,expected)


    def testTheta(self):
        """test theta"""
	profile_base = {'D3S1358':(('15',113),('17',89)), \
		'VWA':(('14',34),('17',100)), \
		'D16S539':(('11',119),('13',70)), \
		'D2S1338':(('24',40),('25',48)), \
		'D8S1179':(('11',25),('13',139)), \
		'D21S11':(('30',105),('31.2',42)), \
		'D18S51':(('14',67),('15',57)), \
		'D19S433':(('14',131),('15.2',10)), \
		'TH01':(('9',55),('9.3',140)), \
		'FGA':(('21',71),('21', 71))}

	profile = {}
	for i in profile_base:
		p = profile_base[i]
		profile[i] = ((p[0][0],float(p[0][1])/400),(p[1][0],float(p[1][1])/400))

	result = strmarker.calc_profile_match_probability(profile,0.0)
	expected = 7.58e-14
	#print "result",result,expected
	self.assertAlmostEqual(result*1e8,expected*1e8)

	result = strmarker.calc_profile_match_probability(profile,0.01)
	expected = 2.44e-13
	#print "result 0.01",result,expected
	self.assertAlmostEqual(result*1e7,expected*1e7)

	result = strmarker.calc_profile_match_probability(profile,0.03)
	expected = 1.614e-12
	#print "result 0.03",result,expected
	self.assertAlmostEqual(result*1e7,expected*1e7)

	# size bias correction
	for i in profile_base:
		p = profile_base[i]
		if p[0][1] == p[1][1]:
			profile[i] = ((p[0][0],float(p[0][1]+4)/404),(p[1][0],float(p[1][1]+4)/404))
		else:
			profile[i] = ((p[0][0],float(p[0][1]+2)/404),(p[1][0],float(p[1][1]+2)/404))

	result = strmarker.calc_profile_match_probability(profile,0.0)
	expected = 1.4225e-13
	#print "result 0.0",result,expected
	self.assertAlmostEqual(result*1e10,expected*1e10)


    def testX(self):
        """test RMPs"""
	data = [{'name':'FG','count':400,'marker':'D3S1358','alleles':{'15':113,'17':89}}, \
		{'name':'FG','count':400,'marker':'VWA','alleles':{'14':34,'17':100}}, \
		{'name':'FG','count':400,'marker':'D16S539','alleles':{'11':119,'13':70}}, \
		{'name':'FG','count':400,'marker':'D2S1338','alleles':{'24':40,'25':48}}, \
		{'name':'FG','count':400,'marker':'D8S1179','alleles':{'11':25,'13':139}}, \

		{'name':'FG','count':400,'marker':'D21S11','alleles':{'30':105,'31.2':42}}, \
		{'name':'FG','count':400,'marker':'D18S51','alleles':{'14':67,'15':57}}, \
		{'name':'FG','count':400,'marker':'D19S433','alleles':{'14':131,'15.2':10}}, \
		{'name':'FG','count':400,'marker':'TH01','alleles':{'9':55,'9.3':140}}, \
		{'name':'FG','count':400,'marker':'FGA','alleles':{'21':71}}]
	# size bias correction
	for d in data:
		d['count'] += 4
		count = d['count']
		for j in d['alleles']:
			if len(d['alleles']) == 1:
				d['alleles'][j] = float(d['alleles'][j]+4)/count
			else:
				d['alleles'][j] = float(d['alleles'][j]+2)/count
	expected = {'D3S1358':0.1282,'VWA':0.0450,'D16S539':0.1068,'D2S1338':0.0257,'D8S1179':0.0466, \
			'D21S11':0.0577,'D18S51':0.0499,'D19S433':0.0196,'TH01':0.0992,'FGA':0.0345}
	#print data

    def testRMPs(self):
	data = [{'name':'AB','count':400,'marker':'FGA','alleles':self.AB_Cau_FGA_alleles}, \
		{'name':'AB','count':400,'marker':'TH01','alleles':self.AB_Cau_TH01_alleles}, \
		{'name':'AB','count':400,'marker':'D16S539','alleles':self.AB_Cau_D16S539_alleles}]
	result = strmarker.calc_rmps(data,'AB',0,0.0)
	expected = {'FGA':0.036,'TH01':0.094,'D16S539':0.103}
	for i in expected:
		print i
		self.assertAlmostEqual(result[i],expected[i],3)

    def testPoolAlleles(self):
        """pool alleles"""
	# constructed counts (n=100)
	alleles = {'a':1,'b':2,'c':3,'d':4,'e':5,'f':15,'g':30,'h':40}
	self.assertEqual(strmarker.pool_alleles(alleles,0,0),alleles)
	expected = {'other':10,'e':5,'f':15,'g':30,'h':40}
	self.assertEqual(strmarker.pool_alleles(alleles,5,0),expected)

	# constructed frequencies (n=100)
	alleles = {'a':0.01,'b':0.02,'c':0.03,'d':0.04,'e':0.05,'f':0.15,'g':0.30,'h':0.40}
	expected = alleles
	result = strmarker.pool_alleles(alleles,0,100)
	for i in expected:
		self.assertAlmostEqual(result[i],alleles[i])
	expected = {'other':0.10,'e':0.05,'f':0.15,'g':0.30,'h':0.40}
	result = strmarker.pool_alleles(alleles,5,100)
	for i in expected:
		self.assertAlmostEqual(result[i],expected[i])

	# AB Cau D16S539 counts (n=200)
	alleles = {'8':6,'9':43,'10':17,'11':119,'12':137,'13':70,'14':7,'15':1}
	self.assertEqual(strmarker.pool_alleles(alleles,0,0),alleles)
	expected = {'9':43,'10':17,'11':119,'12':137,'13':70,'14':7,'other':7}
	self.assertEqual(strmarker.pool_alleles(alleles,5,0),expected)

	# AB Cau D16S539 frequencies (n=200)
	alleles = self.AB_Cau_D16S539_alleles
	expected = alleles
	result = strmarker.pool_alleles(alleles,0,200)
	for i in expected:
		self.assertAlmostEqual(result[i],expected[i])
	expected = self.fromPercentages({'9':10.75,'10':4.25,'11':29.75,'12':34.25,'13':17.5,'14':1.75,'other':1.75})
	result = strmarker.pool_alleles(alleles,5,200*2)
	for i in expected:
		self.assertAlmostEqual(result[i],expected[i])



if __name__ == "__main__":
	unittest.main()

test/test_strmarker.py
"""
strmarker test module.
"""

import unittest

from strprofiles import strmarker


class exampleTestCase(unittest.TestCase):

    def setUp(self):
	# AB Cau FGA frequencies (n=200)
	self.AB_Cau_FGA_alleles = {'18': 0.015, '19': 0.0625, '20': 0.1625, '20.2': 0.0075, '21': 0.1775, '22': 0.165, '22.2': 0.005, '23': 0.14, '24': 0.1325, '25': 0.1125, '26': 0.015, '27': 0.005}
	# AB Cau TH01 frequencies (n=200)
	self.AB_Cau_TH01_alleles = self.fromPercentages({'6':25.25, '7':16.00, '8':9.25, '9':13.75, '9.3':35.00, 10:0.75})
	# AB Cau D16S539 frequencies (n=200)
	self.AB_Cau_D16S539_alleles = self.fromPercentages({'8':1.5,'9':10.75,'10':4.25,'11':29.75,'12':34.25,'13':17.5,'14':1.75,'15':0.25})


    def tearDown(self):
        pass


    def testMarkerRandomMatchProbability(self):
        """marker random match probability"""

	# AB Cau FGA frequencies (n=200)
	alleles = self.AB_Cau_FGA_alleles
	expected = 0.036
	result = strmarker.calc_marker_rmp(alleles,0.0)
        self.assertAlmostEqual(result,expected,3)

	# AB Cau TH01 frequencies (n=200)
	alleles = self.AB_Cau_TH01_alleles
	expected = 0.094
	result = strmarker.calc_marker_rmp(alleles,0.0)
        self.assertAlmostEqual(result,expected,3)

	# AB Cau D16S539 frequencies (n=200)
	alleles = self.AB_Cau_D16S539_alleles
	expected = 0.103
	result = strmarker.calc_marker_rmp(alleles,0.0)
        self.assertAlmostEqual(result,expected,3)


    def fromPercentages(self,alleles):
        for i in alleles:
            alleles[i] /= 100.0
        return alleles


    def testGetModalProfile(self):
        """get modal profile"""
	item = {'name':'AB','marker':'VWA','alleles':{'5':0.94,'6':0.03,'7':0.02,'8':0.01}}
	data = [{'name':'AB','count':400,'marker':'FGA','alleles':self.AB_Cau_FGA_alleles}, \
		{'name':'AB','count':400,'marker':'TH01','alleles':self.AB_Cau_TH01_alleles}, \
		{'name':'AB','count':400,'marker':'D16S539','alleles':self.AB_Cau_D16S539_alleles},item]
	expected = {'FGA':(('21',0.1775),('22',0.165)), 'TH01':(('9.3',0.35),('6',0.2525)),'D16S539':(('12',0.3425),('11',0.2975)),'VWA':(('5',0.94),('5',0.94))}
	profile = expected
	result = strmarker.get_modal_profile(data,'AB')
	for i in expected:
		self.assertEqual(result[i][0][0],expected[i][0][0])
		self.assertEqual(result[i][1][0],expected[i][1][0])
		self.assertAlmostEqual(result[i][0][1],expected[i][0][1])
		self.assertAlmostEqual(result[i][1][1],expected[i][1][1])
	expected = 2*0.1775*0.165 * 2*0.35*0.2525 * 2*0.3425*0.2975 * 0.94*0.94
	result = strmarker.calc_profile_match_probability(profile,0.0)
	self.assertAlmostEqual(result,expected)


    def testTheta(self):
        """test theta"""
	profile_base = {'D3S1358':(('15',113),('17',89)), \
		'VWA':(('14',34),('17',100)), \
		'D16S539':(('11',119),('13',70)), \
		'D2S1338':(('24',40),('25',48)), \
		'D8S1179':(('11',25),('13',139)), \
		'D21S11':(('30',105),('31.2',42)), \
		'D18S51':(('14',67),('15',57)), \
		'D19S433':(('14',131),('15.2',10)), \
		'TH01':(('9',55),('9.3',140)), \
		'FGA':(('21',71),('21', 71))}

	profile = {}
	for i in profile_base:
		p = profile_base[i]
		profile[i] = ((p[0][0],float(p[0][1])/400),(p[1][0],float(p[1][1])/400))

	result = strmarker.calc_profile_match_probability(profile,0.0)
	expected = 7.58e-14
	#print "result",result,expected
	self.assertAlmostEqual(result,expected,16)

	result = strmarker.calc_profile_match_probability(profile,0.01)
	expected = 2.44e-13
	#print "result 0.01",result,expected
	self.assertAlmostEqual(result,expected,15)

	result = strmarker.calc_profile_match_probability(profile,0.03)
	expected = 1.614e-12
	#print "result 0.03",result,expected
	self.assertAlmostEqual(result,expected,14)

	# size bias correction
	for i in profile_base:
		p = profile_base[i]
		if p[0][1] == p[1][1]:
			profile[i] = ((p[0][0],float(p[0][1]+4)/404),(p[1][0],float(p[1][1]+4)/404))
		else:
			profile[i] = ((p[0][0],float(p[0][1]+2)/404),(p[1][0],float(p[1][1]+2)/404))

	result = strmarker.calc_profile_match_probability(profile,0.0)
	expected = 1.4225e-13
	#print "result 0.0",result,expected
	self.assertAlmostEqual(result,expected,17)


    def testRMPs(self):
        """test RMPs"""
	data = [{'name':'AB','count':400,'marker':'FGA','alleles':self.AB_Cau_FGA_alleles}, \
		{'name':'AB','count':400,'marker':'TH01','alleles':self.AB_Cau_TH01_alleles}, \
		{'name':'AB','count':400,'marker':'D16S539','alleles':self.AB_Cau_D16S539_alleles}]
	result = strmarker.calc_rmps(data,'AB',0,0.0)
	expected = {'FGA':0.036,'TH01':0.094,'D16S539':0.103}
	for i in expected:
		print i
		self.assertAlmostEqual(result[i],expected[i],3)


    def testPoolAlleles(self):
        """test pool alleles"""
	# constructed counts (n=100)
	alleles = {'a':1,'b':2,'c':3,'d':4,'e':5,'f':15,'g':30,'h':40}
	self.assertEqual(strmarker.pool_alleles(alleles,0,0),alleles)
	expected = {'other':10,'e':5,'f':15,'g':30,'h':40}
	self.assertEqual(strmarker.pool_alleles(alleles,5,0),expected)

	# constructed frequencies (n=100)
	alleles = {'a':0.01,'b':0.02,'c':0.03,'d':0.04,'e':0.05,'f':0.15,'g':0.30,'h':0.40}
	expected = alleles
	result = strmarker.pool_alleles(alleles,0,100)
	for i in expected:
		self.assertAlmostEqual(result[i],alleles[i])
	expected = {'other':0.10,'e':0.05,'f':0.15,'g':0.30,'h':0.40}
	result = strmarker.pool_alleles(alleles,5,100)
	for i in expected:
		self.assertAlmostEqual(result[i],expected[i])

	# AB Cau D16S539 counts (n=200)
	alleles = {'8':6,'9':43,'10':17,'11':119,'12':137,'13':70,'14':7,'15':1}
	self.assertEqual(strmarker.pool_alleles(alleles,0,0),alleles)
	expected = {'9':43,'10':17,'11':119,'12':137,'13':70,'14':7,'other':7}
	self.assertEqual(strmarker.pool_alleles(alleles,5,0),expected)

	# AB Cau D16S539 frequencies (n=200)
	alleles = self.AB_Cau_D16S539_alleles
	expected = alleles
	result = strmarker.pool_alleles(alleles,0,200)
	for i in expected:
		self.assertAlmostEqual(result[i],expected[i])
	expected = self.fromPercentages({'9':10.75,'10':4.25,'11':29.75,'12':34.25,'13':17.5,'14':1.75,'other':1.75})
	result = strmarker.pool_alleles(alleles,5,200*2)
	for i in expected:
		self.assertAlmostEqual(result[i],expected[i])


if __name__ == "__main__":
	unittest.main()

strprofiles/strmarker.py
"""
strmarker

Utilities for STR (short tandem repeat) genetic markers and allele frequency data.
"""

from operator import itemgetter


SGM_PLUS_MARKERS = ['FGA','TH01','VWA','D2S1338','D3S1358','D8S1179','D16S539','D18S51','D19S433','D21S11']

def calc_marker_rmp(alleles,theta):
	"""
	Calculate the random match probability for at a genetic marker, given the allele frequencies
	for that marker. The Balding and Nichols formulae are used.
	Keyword arguments:
	alleles -- a dict of allele frequencies in the form {'value':frequency,...}
	theta -- the population subdivision coefficient, used to correct for subdivided populations

	"""
	rmp = 0.0;
	denom = (1+theta)*(1+2*theta)
	for i in alleles:
		for j in alleles:
			if i==j:
				p = (2*theta+(1-theta)*alleles[i])*(3*theta+(1-theta)*alleles[i])/denom
				rmp += p*p
			else:
				p = (theta+(1-theta)*alleles[i])*(theta+(1-theta)*alleles[j])/denom
				rmp += 2*p*p
	return rmp


def pool_alleles(alleles,cutoff,count):
	"""
	Pool low frequency alleles together

	Keyword arguments:
	alleles -- a dict of allele frequencies in the form {'value':frequency,...}
	cutoff -- the minimum size of a frequency bin, items with a frequency lower than this will be pooled
	count -- the size of the sample from which the frequencies were derived

	"""
	items = alleles.items()
	items.sort(key = itemgetter(1))
	limit = float(cutoff)
	if count != 0:
		limit = float(cutoff)/float(count)
	otherCount = 0.0
	ret = {}
	for i in items:
		if i[1]<limit or otherCount<limit:
			otherCount += i[1]
		else:
			ret[i[0]] = i[1]
	if otherCount != 0:
		ret['other'] = otherCount
	return ret


def calc_rmps(data,name,cutoff,theta):
	"""
	Calculate the random match probabilities for each genetic marker for the named sample set

	Keyword arguments:
	data -- the allele frequency data
	name -- the name of the sample set to be used
	cutoff -- the minimum size of a frequency bin, items with a frequency lower than this will be pooled
	theta -- the population subdivision coefficient, used to correct for subdivided populations

	"""
	ret = {}
	rmp = 1.0
	for d in data:
		if d['name']==name and d['marker'] in SGM_PLUS_MARKERS:
			count = d['count']
			# note, count doubled since two allele values per person in sample
			alleles = pool_alleles(d['alleles'],cutoff,2*count)
			p = calc_marker_rmp(alleles,theta)
			ret[d['marker']] = p
			rmp *= p
	ret['count'] = count
	ret['combined'] = rmp
	ret['reciprocal'] = 1.0/rmp
	return ret


def get_modal_profile(data,name):
	"""
	Find the modal profile for in the named sample set.

	"""
	profile = {}
	for d in data:
		if d['name']==name and d['marker'] in SGM_PLUS_MARKERS:
			items = d['alleles'].items()
			items.sort(key = itemgetter(1))
			items.reverse()
			p = items[0][1]
			q = items[1][1]
			if p*p > 2*p*q:
				profile[d['marker']]  = ((items[0][0],p),(items[0][0],p))
			else:
				profile[d['marker']]  = ((items[0][0],p),(items[1][0],q))
	return profile


def calc_profile_match_probability(profile,theta):
	"""
	Calculate the probability that a random individual matches the given profile

	Keyword arguments:
	profile -- a 2D dict of the data to be formatted
	theta -- the population subdivision coefficient, used to correct for subdivided populations

	"""
	pmp = 1.0
	for marker in profile:
		m = profile[marker]
		p = m[0][1]
		q = m[1][1]
		denom = (1+theta)*(1+2*theta)
		if m[0][0] == m[1][0]:
			mp = (2*theta+(1-theta)*p)*(3*theta+(1-theta)*p)/denom
		else:
			mp = 2*(theta+(1-theta)*p)*(theta+(1-theta)*q)/denom
		pmp *= mp
		#print marker,round(mp,4)
	return pmp




strprofiles/sgm.py
#!/usr/bin/env python
#coding=utf-8
#file: sgm.py

"""
sgm
"""

import strmarker
import csv
import string
from jinja2 import Template
from collections import defaultdict
from optparse import OptionParser


def read_csv(filename,prefix,normalizer):
	"""
	Read genetic marker allele frequency data from a .csv file

	Keyword arguments:
	filename -- the file to be read
	normalizer -- used to normalize the data, use 100 if the data are percentages, 1 if they are
	fractions in the range [0.0,1.0]

	Format of file is:
	First row - each cell is a SGM Marker name, cell A1 blank
	Second row - each cell is sample size and sample name (eg "302 Cau"), cell B1 ignored
	First column - (starting at C1) allele value (eg "11"), cell A1 blank, cell B1 is column header "Allele"
	Main table - from C2 - allele frequency value for given Marker, Sample and allele

	,"CSF1PO","CSF1PO","CSF1PO","FGA","FGA","FGA",...
	"Allele","302 Cau","258 AA","140 His","302 Cau","258 AA","140 His",...
	7,,0.05253,0.02143,,,,...

	"""
	data = []
	reader = csv.reader(open(filename, "rb"),'excel')

	# read in the first row, which contains the SMG Marker names
	row = reader.next()
	for i in range(1,len(row)-1):
		data.append({'marker':row[i],'alleles':{}})

	# read in the second row, which contains the sample names
	row = reader.next()
	for i in range(1,len(row)-1):
		vals = string.split(row[i])
		if len(vals) == 1:
			data[i-1]['name'] = prefix + vals[0]
		else:
			data[i-1]['count'] = int(vals[0])
			data[i-1]['name'] = prefix + vals[1]

	# read in the allele frequencies
	try:
		while 1:
			row = reader.next()
			for i in range(1,len(row)-1):
				if row[i] != '':
					data[i-1]['alleles'][row[0]] = float(row[i])/normalizer
	except StopIteration:
		pass
	return data


def rmp_table_text(data,caption,rowheaders,colheaders):
	"""
	Format data into a text table formatted using whitespace

	Keyword arguments:
	data -- a 2D dict of the data to be formatted
	caption -- the table's caption
	rowheaders -- the table's row headings
	colheaders -- the table's column headings

	"""
	t = Template('\n{{caption}}\n'
		'{{"%8s"|format("")}} '
		'{% for colheader in colheaders %}'
			'{{"%8s"|format(colheader)}} '
		'{% endfor %}\n'
		'{{"%8s"|format("Count")}} '
		'{% for colheader in colheaders %}'
			'{{"%8d"|format(data[colheader]["count"])|escape}} '
		'{% endfor %}\n'
		'{% for rowheader in rowheaders %}'
			'{{"%8s"|format(rowheader)}} '
			'{% for colheader in colheaders %}'
				'{{"   %1.3f"|format(data[colheader][rowheader])|escape}} '
			'{% endfor %}\n'
		'{% endfor %}'
		'{{"%8s"|format("Combined")}} '
		'{% for colheader in colheaders %}'
			'{{"%1.2e"|format(data[colheader]["combined"])|escape}} '
		'{% endfor %}\n'
		'{{"%8s"|format("Recip")}} '
		'{% for colheader in colheaders %}'
			'{{"%1.2e"|format(data[colheader]["reciprocal"])|escape}} '
		'{% endfor %}')
	return t.render(data=data, caption=caption, rowheaders=rowheaders, colheaders=colheaders)


def pmp_table_html(data,caption,rowheaders,colheaders):
	"""
	Format data into an HTML table

	Keyword arguments:
	data -- a 2D dict of the data to be formatted
	caption -- the table's caption
	rowheaders -- the table's row headings
	colheaders -- the table's column headings

	"""
	t = Template('<html>\n<table>\n'
		'<caption>{{caption}}</caption>\n'
		'<thead>\n'
		'<tr> '
		'<th></th> '
		'{% for colheader in colheaders %}'
			'<th>{{colheader|replace(" ", "<br />")}}</th> '
		'{% endfor %}'
		'</tr>\n'
		'</thead>\n'
		'<tbody>\n'
		'{% for rowheader in rowheaders %}'
			'<tr> '
			'<th>{{rowheader}}</th> '
			'{% for colheader in colheaders %}'
				'<td>{{"%1.2e"|format(data[colheader][rowheader])|escape}}</td> '
			'{% endfor %}'
			'</tr>\n'
		'{% endfor %}'
		'</tbody>\n'
		'</table>\n</html>\n')
	return t.render(data=data, caption=caption, rowheaders=rowheaders, colheaders=colheaders)


def pmp_table_text(data,caption,rowheaders,colheaders):
	"""
	Format data into a text table formatted using whitespace

	Keyword arguments:
	data -- a 2D dict of the data to be formatted
	caption -- the table's caption
	rowheaders -- the table's row headings
	colheaders -- the table's column headings

	"""
	t = Template('\n{{caption}}\n'
		'{{"%8s"|format("")}} '
		'{% for colheader in colheaders %}'
			'{{"%8s"|format(colheader)}} '
		'{% endfor %}\n'
		'{% for rowheader in rowheaders %}'
			'{{"%8s"|format(rowheader)}} '
			'{% for colheader in colheaders %}'
				'{{"%1.2e"|format(data[colheader][rowheader])|escape}} '
			'{% endfor %}\n'
		'{% endfor %}')
	return t.render(data=data, caption=caption, rowheaders=rowheaders, colheaders=colheaders)


def rmp_table_html(data,caption,rowheaders,colheaders):
	"""
	Format data into an HTML table

	Keyword arguments:
	data -- a 2D dict of the data to be formatted
	caption -- the table's caption
	rowheaders -- the table's row headings
	colheaders -- the table's column headings

	"""
	t = Template('<html>\n<table>\n'
		'<caption>{{caption}}</caption>\n'
		'<thead>\n'
		'<tr> '
		'<th></th> '
		'{% for colheader in colheaders %}'
			'<th>{{colheader|replace(" ", "<br />")}}</th> '
		'{% endfor %}'
		'</tr>\n'
		'<tr> '
		'<th>Count</th> '
		'{% for colheader in colheaders %}'
			'<th>{{data[colheader]["count"]|escape}}</th> '
		'{% endfor %}\n'
		'</tr>\n'
		'</thead>\n'
		'<tbody>\n'
		'{% for rowheader in rowheaders %}'
			'<tr> '
			'<th>{{rowheader}}</th> '
			'{% for colheader in colheaders %}'
				'<td>{{" %1.3f"|format(data[colheader][rowheader])|escape}}</td> '
			'{% endfor %}'
			'</tr>\n'
		'{% endfor %}'
		'</tbody>\n'
		'<tr> '
		'<th>Combined</th> '
		'{% for colheader in colheaders %}'
			'<td>{{"%1.2e"|format(data[colheader]["combined"])|escape}}</td> '
		'{% endfor %}'
		'</tr>\n'
		'<tr> '
		'<th>Reciprocal</th> '
		'{% for colheader in colheaders %}'
			'<td>{{"%1.2e"|format(data[colheader]["reciprocal"])|escape}}</td> '
		'{% endfor %}'
		'</tr>\n'
		'<tfoot>\n'
		'</tfoot>\n'
		'</table>\n</html>\n')
	return t.render(data=data, caption=caption, rowheaders=rowheaders, colheaders=colheaders)


def tabulate_rmps(data,samples,format,caption,cutoff,theta):
	"""
	Calculate the random match probabilities and format them into a table
	"""
	table = defaultdict(dict)
	columnHeaders = []
	for sample in samples:
		columnHeaders.append(sample)
		table[sample] = strmarker.calc_rmps(data,sample,cutoff,theta)
	if format == "html":
		return rmp_table_html(table,caption,strmarker.SGM_PLUS_MARKERS,columnHeaders)
	else:
		return rmp_table_text(table,caption,strmarker.SGM_PLUS_MARKERS,columnHeaders)


def tabulate_pmps(data,samples,format,caption):
	"""
	Calculate the profile match probabilities for the modal profile and format them into a table
	"""
	table = defaultdict(dict)
	columnHeaders = []
	for sample in samples:
		columnHeaders.append(sample)
		profile = strmarker.get_modal_profile(data,sample)
		table[sample]['0.0'] = 1.0/strmarker.calc_profile_match_probability(profile,0.0)
		table[sample]['0.01'] = 1.0/strmarker.calc_profile_match_probability(profile,0.01)
		table[sample]['0.03'] = 1.0/strmarker.calc_profile_match_probability(profile,0.03)
	if format == "html":
		return pmp_table_html(table,caption,['0.0','0.01','0.03'],columnHeaders)
	else:
		return pmp_table_text(table,caption,['0.0','0.01','0.03'],columnHeaders)


def main():
	"""
	Read in the allele frequency data
	Print out tables of random match probabilities.
	Print out the table of profile match probabilities for the modal profile.

	"""

	parser = OptionParser()
	parser.add_option("-t", action="store_true", dest="text_format", default=False, help="use text format tables")
 	parser.add_option("-v", action="store_true", dest="verbose", default=False, help="print status messages to stdout")
	(options,args) = parser.parse_args()
	if options.text_format:
		format = "text"
	else:
		format = "html"

	# read in the NIST/JSF allele frequency data
	dataJFS = read_csv("../data/JFS2003IDresults.csv","JSF ",1)

	# read in the NIST/JSF allele frequency data
	dataAB = read_csv("../data/ABresults.csv","AB ",100)
	data = dataJFS + dataAB

	samples = ['JSF AA','JSF Cau','JSF His','AB AA','AB Cau']

	print tabulate_rmps(data,samples,format,"Raw Probability of Identity values",0,0.0)
	print tabulate_rmps(data,samples,format,"Rare alleles pooled",5,0.0)
	print tabulate_rmps(data,samples,format,"Theta = 0.01",5,0.01)
	print tabulate_rmps(data,samples,format,"Theta = 0.03",5,0.03)
	print tabulate_pmps(data,samples,format,"Modal Man")


if __name__ == "__main__":
	main()